dipy/nn/synb0.py

```diff
@@ -0,0 +1,417 @@
+                                                            use_bias=False)
+                self.instnorm = tfa.layers.InstanceNormalization()
+                self.activation = tf.keras.layers.LeakyReLU(0.01)
+
+            def call(self, input):
+                x = self.conv3d(input)
+                x = self.instnorm(x)
+                x = self.activation(x)
+
+                return x
+
+    def fetch_default_weights(self, idx):
+        r"""
+        Load the model pre-training weights to use for the fitting.
+
+        Parameters
+        ----------
+        idx : int
+            The idx of the default weights. It can be from 0~4.
+        """
+        fetch_model_weights_path = get_fnames('synb0_default_weights')
+        print('fetched ' + fetch_model_weights_path[idx])
+        self.load_model_weights(fetch_model_weights_path[idx])
+
+    def load_model_weights(self, weights_path):
+        r"""
+        Load the custom pre-training weights to use for the fitting.
+
+        Parameters
+        ----------
+        weights_path : str
+            Path to the file containing the weights (hdf5, saved by tensorflow)
+        """
+        try:
+            self.model.load_weights(weights_path)
+        except ValueError:
+            raise ValueError('Expected input for the provided model weights \
+                             do not match the declared model')
+
+    def __predict(self, x_test):
+        r"""
+        Internal prediction function
+
+        Parameters
+        ----------
+        x_test : np.ndarray (batch, 80, 80, 96, 2)
+            Image should match the required shape of the model.
+
+        Returns
+        -------
+        np.ndarray (...) or (batch, ...)
+            Reconstructed b-inf image(s)
+        """
+
+        return self.model.predict(x_test)
+
+    def __normalize(self, image, max_img, min_img):
+        r"""
+        Internal normalization function
+
+        Parameters
+        ----------
+        image : np.ndarray
+
+        Returns
+        -------
+        np.ndarray
+            Normalized image from range -1 to 1
+        """
+        return ((image - min_img)/(max_img - min_img))*2-1
+
+    def __unnormalize(self, image, max_img, min_img):
+        r"""
+        Internal unnormalization function
+
+        Parameters
+        ----------
+        image : np.ndarray
+
+        Returns
+        -------
+        np.ndarray
+            unnormalized image from range min_img to max_img
+        """
+        return (image+1)/2*(max_img-min_img) + min_img
+
+    def predict(self, b0, T1, batch_size=None, average=True):
+        r"""
+        Wrapper function to faciliate prediction of larger dataset.
+        The function will scale the data to meet the required shape of image.
+        Note that the b0 and T1 image should have the same shape
+
+        Parameters
+        ----------
+        b0 : np.ndarray (batch, 77, 91, 77) or (77, 91, 77)
```

Comment:
This shape is a whole image? Does it mean it won't work with the vast majority of images out there?

What is the solution when someone has an image like (100,90,100?) Does it have to be cropped (and loose tissue) or resample to an arbitrary voxel size?

Or is the expected preprocessing to rigidly register the volume to a template with these dimensions?

Reply:
I am not sure how the shape is created. This model is intended to be a part of the pipeline, hence assumes it gets the intended input shape.

Reply:
Moving forward we have to think about these constraints.